### PR TITLE
Let use an arbitrary precision library in the AkeneoMeasuresBundle

### PR DESCRIPTION
**Pull request by gplanchat**
**Description (for Contributor and Core Developer)**

See #5137

By replacing the type hints of concrete class by an interface, it is possible to implement a high-precision measures conversion

**Definition Of Done (for Core Developer only)**

| Q | A |
| --- | --- |
| Added Specs |  |
| Added Behats |  |
| Changelog updated |  |
| Review and 2 GTM |  |
| Micro Demo to the PO (Story only) |  |
| Migration script |  |
|  Tech Doc |  |

### CODE REVIEW

File: src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * Convert a value in a standard unit to a final unit
+     * @param string $finalUnit Final unit for value
+     * @param double $value     Value to convert
+     *
+     * @throws UnknownOperatorException
+     * @throws UnknownMeasureException
+     * @return double
+     *
+     */
+    public function convertStandardToResult($finalUnit, $value);
```
Comment (nidup) [inline src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php]: this method should not be in the interface and should be protected in the concret implem, it's never used except from the concret implementation itself (from convert method)

File: src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+{
+    /**
+     * Set a family for the converter
+     * @param string $family
+     *
+     * @throws UnknownFamilyMeasureException
+     * @return MeasureConverterInterface
+     *
+     */
+    public function setFamily($family);
```
Comment (nidup) [inline src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php]: This method should be removed, I checked in the code and each time we call convert() or convertBaseToStandard() we call the setFamily just before. I'd prefer that we pass the family as first argument of  convert() and convertBaseToStandard() to make this code more readable and solid.

File: src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Convert a value from a base measure to a final measure
+     * @param string $baseUnit  Base unit for value
+     * @param string $finalUnit Result unit for value
+     * @param double $value     Value to convert
+     *
+     * @return double
+     */
+    public function convert($baseUnit, $finalUnit, $value);
```
Comment (nidup) [inline src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php]: Should be `public function convert($family, $baseUnit, $finalUnit, $value);`

File: src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+     * Convert a value in a base unit to the standard unit
+     * @param string $baseUnit Base unit for value
+     * @param double $value    Value to convert
+     *
+     * @throws UnknownOperatorException
+     * @throws UnknownMeasureException
+     * @return double
+     *
+     */
+    public function convertBaseToStandard($baseUnit, $value);
```
Comment (nidup) [inline src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php]: Should be `public function convertBaseToStandard($family, $baseUnit, $value);`

File: src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace Akeneo\Bundle\MeasureBundle\Convert;
+
+use Akeneo\Bundle\MeasureBundle\Exception\UnknownFamilyMeasureException;
+use Akeneo\Bundle\MeasureBundle\Exception\UnknownMeasureException;
+use Akeneo\Bundle\MeasureBundle\Exception\UnknownOperatorException;
+
+/**
+ * Aims to convert measures
```
Comment (nidup) [inline src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php]: Could be replaced by `Convert from a measure family unit to another unit of the same family, for instance 1 KILOGRAM to 1000 GRAM for the family Weight`

File: src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php (hunk 3 above)
Comment (nidup) [inline src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php]: Doulbe thinking of this API, the argument naming is not understandable and the order seems really not natural, would prefer `convert($family, $value, $fromUnit, $toUnit);`

File: src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php (hunk 4 above)
Comment (nidup) [inline src/Akeneo/Bundle/MeasureBundle/Convert/MeasureConverterInterface.php]: Same weird naming for method and arguments, would prefer `convertToStandardUnit($family, $value, $fromUnit)`

**Comment by nidup** [comment]
Thx for your PR, reviewing it I think about couple of things,

1) as we extract a new interface, we should rework a bit the naming, args and contract to make it more clear (this code is pretty old and legacy making an ugly API)
2) we could extract a Akeneo/Component/Measure from this bundle, this requires a bit more work especially from migration point of view but this could be done in a second time / second PR

Could you fix my comments that will allow us to fulfill the 1/ step? (once merged, I'll take care of documenting related BC breaks). Thx :tada:

**Comment by gplanchat** [comment]
Changing the API will silently break a lot of related code in the PIM and in custom bundles, it is way out of the initial scope of this PR :)

**Comment by aRn0D** [comment]
Hi, Grégory @gplanchat thank you for your contribution! I will close this PR because what we want, require more than the scope of that PR.

**Pull request closed without merging**